src/transformers/cache_utils.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, Tuple, TypeVar
+import torch
+
+T = TypeVar("T")
+
+
+class Cache(ABC):
+    def __init__(self) -> None:
+        self.cache: Dict[int, Tuple[torch.Tensor]] = {}
+        self.layer_idx = 0
+
+    @abstractmethod
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        pass
+
+    @abstractmethod
+    def update(self, key_states, value_states) -> None:
+        pass
+
+    def __getitem__(self, index: int):
+        return self.cache[self.layer_idx][index]
+
+    def set_layer_idx(self, layer_idx: int) -> None:
+        self.layer_idx = layer_idx
```

Comment:
We want our `Cache` instance to work both for `past_key_values` and `past_key_value`. This is one of the approaches to make that work - when we want to get the KV cache for a specific layer we first set the layer idx and then slicing works as if we have `past_key_value`.

Other alternatives might require creating some instance for each of the (e.g. 32) layers and then having two-way commmunication between them, which I thought was suboptimal. But I'm not 100% a fan of this approach either, as you now can't use indexing to select layers like you might expect.

Reply:
Regardless of the class structure we decide to keep, it would be really really great if we could keep the same indexing capabilities as the previous `past_key_values` structure. i.e. the old `past_key_values[a, b, c, d]` would match the new Cache-based `past_key_values[a, b, c, d]`

This would make the transition on the `generation` side much easier -- advanced features like assisted generation or fixed-sized caches rely on indexing, and it would make the transition process much easier 💛

Reply:
Makes sense. I'm not sure how `past_key_values` is indexed from the generation mixin as I've only looked at that code a few times, but I'll try to ensure 1-1 parity for indexing. I think this should be feasible, I already have ideas 😄

Reply:
I would actually not add such a function and also not add a `self.layer_index` attribute to the cache class. Instead we can pass the index directly to `update(...)` so that we can keep the class as easy-to-understand as possible. This has the following advantages:
- The user sees less logic related to the cache in the modeling file (only one `update(...)`) function and instead of `set_layer_index(...)` and `update(...)` => easier to understand
- We keep the cache object more light-weight
- It doesn't really make sense for a cache to be have a layer_index attribute IMO

Reply:
I can remove the `layer_idx` from `Cache`, but the layer idx will have to be passed all the way down to the `...Attention` classes instead where `update` is called then, either via `forward` or `__init__`. I was trying to avoid this.
Or otherwise the `Cache` needs to maintain a list of smaller `KVCache` (or some other name) instances, one for each layer, and pass those down as `past_key_value` instead. That would also work, but it's not exactly lightweight either to have two separate classes and e.g. 33 instances just for one cache.

I do agree with your 3 advantages, though

Reply:
Oh, I didn't realise you already did this. Yeah, I'm totally fine with that then.

Reply:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.